### PR TITLE
Resilience operators: Retry

### PR DESCRIPTION
**Pull request by gcotelli**
Add a first version of `Retry`

### CODE REVIEW

File: docs/Resilience.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+Many faults are transient and may self-correct after a short time. So this helps in auto-recovery of transient errors.
+
+### Usage
+```smalltalk
+Retry
+ value: [ "operation to execute" ]
+ configuredBy: [:retry | "configuration options" ]
+```
+
+The block to be evaluated can optionally receive the current number of attempt in a block argument.
```
Comment (mtabacman) [inline docs/Resilience.md]:
```suggestion
The block to be evaluated can optionally receive the current attempt  number in a block argument. 
```

Comment (mtabacman) [inline docs/Resilience.md]: Can we get an example for that?

File: docs/Resilience.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+```smalltalk
+Retry
+ value: [ "operation to execute" ]
+ configuredBy: [:retry | "configuration options" ]
+```
+
+The block to be evaluated can optionally receive the current number of attempt in a block argument. 
+
+### Configuration Options
+This options are sent to the `retry` instance provided in the second block.
```
Comment (mtabacman) [inline docs/Resilience.md]:
```suggestion
These options are sent to the `retry` instance provided in the second block.
```

File: docs/Resilience.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+-	`upTo: retryCount` : The maximum number of retries. Defaults to 2.
+-	`every: duration` : Wait a time duration between retry attempts. By default don't wait.
```
Comment (mtabacman) [inline docs/Resilience.md]:
```suggestion
-	`every: duration` : Wait time duration between retry attempts. By default don't wait.
```

File: docs/Resilience.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+- `backoffExponentiallyWithTimeSlot: duration` : Wait a time duration between retry attempts determined by using the [exponential backoff algorithm](https://en.wikipedia.org/wiki/Exponential_backoff) with a `duration` time slot.
```
Comment (mtabacman) [inline docs/Resilience.md]:
```suggestion
- `backoffExponentiallyWithTimeSlot: duration` : Wait time duration between retry attempts determined by using the [exponential backoff algorithm](https://en.wikipedia.org/wiki/Exponential_backoff) with a `duration` time slot.
```

File: docs/Resilience.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+- `when: condition` : Evaluate the condition with the current execution result, if met retry the execution.
+- `when: condition evaluating: action` : Like the previous one but `action` will be evaluated when a retry is made due to the condition. `action` is a block receiving as optional arguments the `attemptNumber` and the current `result`.
```
Comment (mtabacman) [inline docs/Resilience.md]: Maybe this requires an example here.

File: docs/Resilience.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+-	`on: exceptionSelector` : Retry the execution if an exception handled by `exceptionSelector` is raised.
+- `on: exceptionSelector evaluating: action` : Like the previous one but `action` will be evaluated when a retry is made due to a handled exception raised. `action` is a block receiving as optional arguments the `attemptNumber` and the `exception` raised.
```
Comment (mtabacman) [inline docs/Resilience.md]: Some example here will clarify usage.

File: docs/Resilience.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+- `ignore: exceptionSelector` : Scope the previous method ignoring the exceptions that can be handled by `exceptionSelector`.
```
Comment (mtabacman) [inline docs/Resilience.md]: I don't know that that means. Also, an example should be added.

File: source/Hyperspace-Model-Tests/RetryTest.class.st [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+				self signalExpectedError
+				]
+			configuredBy: [ :retry | 
+				retry
+					onException;
+					backoffExponentiallyWithTimeSlot: 50 milliSeconds
+				]
+		].
+
+	self assert: count equals: 3
```
Comment (mtabacman) [inline source/Hyperspace-Model-Tests/RetryTest.class.st:130]: From reading the test it's not clear why this is how it should happen.

File: source/Hyperspace-Model/Retry.class.st [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+	self on: Exception
+]
+
+{ #category : #configuring }
+Retry >> upTo: aRetryCount [
+
+	AssertionChecker enforce: [ aRetryCount strictlyPositive ]
+		because: 'The maximum number of retries must be strictly positive.'.
+	maximumNumberOfAttempts := aRetryCount + 1
```
Comment (mtabacman) [inline source/Hyperspace-Model/Retry.class.st:136]: There's a +1 here and a -1 when using it. Shoudn't they somehow cancel each other out in the code and not be required? Maybe just renaming the variable or something...

**Review by mtabacman — CHANGES_REQUESTED** [review]

File: source/Hyperspace-Model/Retry.class.st (hunk 9 above)
Comment (gcotelli) [reply source/Hyperspace-Model/Retry.class.st:136]: The variable is used in 4 places.. 2 are decrementing/incrementing and the others not... so it's the same

File: source/Hyperspace-Model/Retry.class.st [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+]
+
+{ #category : #configuring }
+Retry >> on: anException [
+
+	exceptionToRetry := anException
+]
+
+{ #category : #configuring }
+Retry >> on: anException evaluating: aDyadicBlock [
```
Comment (fortizpenaloza) [inline source/Hyperspace-Model/Retry.class.st:119]: I was expecting being able to add multiple handlers. The same apply to #on: message. Now I get why it's possible to send #ignore: multiple times. 

Does it makes sense to have more than one handler by adding #on:ignoring: and #on:ignoring:evaluating: messages?

Comment (gcotelli) [reply source/Hyperspace-Model/Retry.class.st:119]: At least for all the use cases we have now nobody is doing something specific for every exception. The block is not intended to change behavior but to produce logging or collect information about each retry made.

**Review by mtabacman — APPROVED** [review]

**Review by fortizpenaloza — APPROVED** [review]

**Pull request merged**